### PR TITLE
Unify Websockets

### LINKED ISSUE #594: Unify Websockets with PubSub logic
**Is your feature request related to a problem? Please describe.**
Currently we have to add a new Websocket endpoint for every websocket feature we want to use

**Describe the solution you'd like**
It would be better to have a single endpoint and subscribe to the channels you want to connect to.

### PR DESCRIPTION
**Pull request by mono424**
Allows communicating in different channels over one socket.
Supports params in channel names like `chat/:streamID`.

Bildschirmfoto 2022-08-04 um 16 00 50

### CODE REVIEW

**Comment by mono424** [comment]
Added more tests to [PR#644](https://github.com/joschahenningsen/TUM-Live/pull/644) as there is a layer which makes testing easier (or even possible).

File: api/chat.go [hunk 1]
```diff
@@ -6,81 +6,72 @@ import (
...
-		m = melody.New()
-	}
+const (
+	ChatRoomName = "chat/:streamID"
+)
 
-	m.HandleConnect(connHandler)
+var routes chatRoutes
+
+//var m *melody.Melody
```
Comment (MatthiasReumann) [inline api/chat.go]: I guess we don't need that out-commented code anymore. Therefore, could you remove it please?

File: api/chat.go [hunk 2]
```diff
@@ -92,6 +83,21 @@ func configGinChatRouter(router *gin.RouterGroup, daoWrapper dao.DaoWrapper) {
...
+	routes = chatRoutes{daoWrapper}
+
+	wsGroup := router.Group("/:streamID")
+	wsGroup.Use(tools.InitStream(daoWrapper))
+	wsGroup.GET("/messages", routes.getMessages)
+	wsGroup.GET("/active-poll", routes.getActivePoll)
+	wsGroup.GET("/users", routes.getUsers)
+	/*wsGroup.GET("/ws", routes.ChatStream)
+	if m == nil {
+		log.Printf("creating melody")
```
Comment (MatthiasReumann) [inline api/chat.go]: Here too.

File: web/ts/socket.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+type MessageHandlerFn = (payload: object) => void;
+
+const RealtimeMessageTypes = {
+    RealtimeMessageTypeSubscribe: "subscribe",
+    RealtimeMessageTypeUnsubscribe: "unsubscribe",
+    RealtimeMessageTypeChannelMessage: "message",
+};
+
+export const realtime = {
+    _debugging: false,
```
Comment (MatthiasReumann) [inline web/ts/socket.ts]: If you'd use TypeScript classes you could utilize TypeScripts `private` visibility keyword vor the `_<name>` attributes and methods.

**Review by MatthiasReumann — COMMENTED** [review]
Works great 💯 Looks also pretty clean code-wise! 

I've left some minor comments. Furthermore, there are some `.DS_Store` files in a bunch of folders which need to be removed. Thanks!

File: api/chat.go (hunk 1 above)
Comment (mono424) [reply api/chat.go:29]: mac ... 👍

**Comment by mono424** [comment]
> Works great 💯 Looks also pretty clean code-wise!
> 
> I've left some minor comments. Furthermore, there are some `.DS_Store` files in a bunch of folders which need to be removed. Thanks!

thanks! yeah mac... will remove them ! 👍

File: web/ts/socket.ts (hunk 3 above)
Comment (mono424) [reply web/ts/socket.ts]: good idea! Refactoring to a singleton ts-class! 👍

**Review by MatthiasReumann — APPROVED** [review]
Looks great 🌟 

One remark: If the changes in the `package-lock.json` weren't intended, revert to the old version please.

**Review by joschahenningsen — APPROVED** [review]
Thanks for all the work! Some remaining nits can be applied in #644

**Pull request merged**